### PR TITLE
Add support Case sensitive identifiers 

### PR DESCRIPTION
**Pull request by Praveen2112**
Relates to #17.
Supersedes #354

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 1]
```diff
@@ -259,6 +259,18 @@ public boolean schemaExists(Session session, CatalogSchemaName schema)
...
+    @Override
+    public String canonicalizeName(Session session, String catalogName, String identifier, boolean delimited)
+    {
+        Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, catalogName);
+
+        if (catalog.isPresent()) {
+            return catalog.get().getMetadata().canonicalize(identifier, delimited);
+        }
+
+        return identifier.toLowerCase(ENGLISH);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java]: Have you run into any usages that fall through to this case?

File: presto-main/src/main/java/io/prestosql/metadata/MetadataUtil.java [hunk 2]
```diff
@@ -146,7 +146,7 @@ public static QualifiedObjectName createQualifiedObjectName(Session session, Nod
     {
         requireNonNull(session, "session is null");
         requireNonNull(name, "name is null");
-        if (name.getParts().size() > 3) {
+        if (name.getOriginalParts().size() > 3) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataUtil.java]: How is this related to "Use canonicalized name for Tables" ?

File: presto-main/src/main/java/io/prestosql/metadata/MetadataUtil.java [hunk 3]
```diff
@@ -134,21 +142,28 @@ public static CatalogSchemaName createCatalogSchemaName(Session session, Node no
         return new CatalogSchemaName(catalogName, schemaName);
     }
 
-    public static QualifiedObjectName createQualifiedObjectName(Session session, Node node, QualifiedName name)
+    public static QualifiedObjectName createQualifiedObjectName(Session session, Node node, QualifiedName name, Metadata metadata)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataUtil.java]: Instead of passing `Metadata`, let's change it to a `NameCanonicalizer` interface with a single method.

File: presto-main/src/main/java/io/prestosql/execution/GrantRolesTask.java [hunk 4]
```diff
@@ -50,8 +49,9 @@ public String getName()
     public ListenableFuture<?> execute(GrantRoles statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
     {
         Session session = stateMachine.getSession();
+        String catalogName = createCatalogName(session, statement);
 
-        Set<String> roles = statement.getRoles().stream().map(role -> role.getValue().toLowerCase(Locale.ENGLISH)).collect(toImmutableSet());
+        Set<String> roles = statement.getRoles().stream().map(role -> metadata.canonicalizeName(session, catalogName, role.getValue(), role.isDelimited())).collect(toImmutableSet());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/execution/GrantRolesTask.java]: Break up the `.map`, `.collect` into separate lines.

File: presto-main/src/main/java/io/prestosql/execution/RevokeRolesTask.java [hunk 5]
```diff
@@ -50,14 +49,14 @@ public String getName()
     public ListenableFuture<?> execute(RevokeRoles statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
     {
         Session session = stateMachine.getSession();
+        String catalog = createCatalogName(session, statement);
 
-        Set<String> roles = statement.getRoles().stream().map(role -> role.getValue().toLowerCase(Locale.ENGLISH)).collect(toImmutableSet());
+        Set<String> roles = statement.getRoles().stream().map(role -> metadata.canonicalizeName(session, catalog, role.getValue(), role.isDelimited())).collect(toImmutableSet());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/execution/RevokeRolesTask.java]: Break up `.map`, `.collect` into separate lines.

File: presto-main/src/main/java/io/prestosql/metadata/Metadata.java [hunk 6]
```diff
@@ -66,12 +66,17 @@
 {
     Set<ConnectorCapabilities> getConnectorCapabilities(Session session, CatalogName catalogName);
 
-    boolean catalogExists(Session session, String catalogName);
+    default boolean catalogExists(Session session, String catalogName)
+    {
+        return getCatalogHandle(session, catalogName).isPresent();
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/Metadata.java]: Why do we need this?

File: presto-main/src/main/java/io/prestosql/metadata/QualifiedTablePrefix.java [hunk 7]
```diff
@@ -44,15 +43,15 @@ public QualifiedTablePrefix(String catalogName)
     public QualifiedTablePrefix(String catalogName, String schemaName)
     {
         this.catalogName = checkCatalogName(catalogName);
-        this.schemaName = Optional.of(checkSchemaName(schemaName));
+        this.schemaName = Optional.of(requireNonNull(schemaName, "schemaName is null"));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/QualifiedTablePrefix.java]: `Optional.of` will validate that the argument is not null, so the `requireNonNull` is unnecessary.

File: presto-main/src/main/java/io/prestosql/metadata/QualifiedTablePrefix.java [hunk 8]
```diff
@@ -44,15 +43,15 @@ public QualifiedTablePrefix(String catalogName)
...
         this.tableName = Optional.empty();
     }
 
     public QualifiedTablePrefix(String catalogName, String schemaName, String tableName)
     {
         this.catalogName = checkCatalogName(catalogName);
-        this.schemaName = Optional.of(checkSchemaName(schemaName));
-        this.tableName = Optional.of(checkTableName(tableName));
+        this.schemaName = Optional.of(requireNonNull(schemaName, "schemaName is null"));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/QualifiedTablePrefix.java]: Same here

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java (hunk 1 above)
Comment (Praveen2112) [reply presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java]: Nope, just a hook till we handle at all the layers
